pkg/job/controller.go

```diff
@@ -0,0 +1,292 @@
+// parent cgroup.
+//
+// Since StopAll is intended for shutdown, it prioritizes completeness over
+// latency and holds the controller's lock for the duration of the process.
+func (c *Controller) StopAll() error {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	if c.shutDown {
+		slog.Info("already shut down")
+		return nil
+	}
+	c.shutDown = true
+
+	errs := []error{}
+	for _, job := range c.jobs {
+		if job.isRunning() {
+			if err := job.stop(); err != nil {
+				errs = append(errs, err)
+			}
+		}
+	}
+	c.wg.Wait() // wait for all jobs to terminate.
+	if err := deleteCgroup(c.telejobCgroup); err != nil {
+		errs = append(errs, err)
+	}
+	if len(errs) > 0 {
+		return errors.Join(errs...)
+	}
+	return nil
+}
+
+// add adds a job to the controller's job map. It is synchronized to ensure safe
+// concurrent access to the job map.
+func (c *Controller) add(id string, job *job) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	c.jobs[id] = job
+}
+
+// get retrieves a job from the controller by ID. It is synchronized to ensure
+// safe concurrent access to the job map. It also verifies that the given owner
+// has access to the requested job.
+func (c *Controller) get(owner, id string) (*job, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	job, ok := c.jobs[id]
+	if !ok {
+		return nil, fmt.Errorf("%w: %q", ErrJobNotFound, id)
+	}
+	if job.owner != owner {
+		return nil, fmt.Errorf("%w: owner %q does not have access to job %q", ErrUnauthorized, owner, id)
+	}
+	return job, nil
+}
+
+// isShutDown reports whether the controller has been shut down. It is
+// synchronized to ensure safe concurrent access to the shutdown status.
+func (c *Controller) isShutDown() bool {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+	return c.shutDown
+}
+
+// newTelejobCgroup creates a new parent cgroup for telejob with the CPU, I/O,
+// and memory resource controllers enabled. It creates the cgroup directory and
+// writes "+cpu +io +memory" to the cgroup.subtree_control file to enable the
+// necessary controllers.
+func newTelejobCgroup(telejobCgroup string) error {
+	err := os.Mkdir(telejobCgroup, 0o750)
+	if err != nil {
+		return fmt.Errorf("cannot create new telejob cgroup %q: %w", telejobCgroup, err)
+	}
+	controlFile := filepath.Join(telejobCgroup, "cgroup.subtree_control")
+	if err := os.WriteFile(controlFile, []byte("+cpu +io +memory"), 0o600); err != nil {
+		return fmt.Errorf("cannot configure cgroup subtree control %q: %w", controlFile, err)
+	}
+	return nil
+}
+
+// newJobCgroup creates a new cgroup for a job with the specified resource
+// limits. The new cgroup is created as a subcgroup under the given parent
+// cgroup. It configures CPU, memory, and I/O limits based on the provided
+// Limits.
+func newJobCgroup(cgroup string, limits Limits) error {
+	if err := os.Mkdir(cgroup, 0o750); err != nil {
+		return fmt.Errorf("cannot create new job cgroup %q: %w", cgroup, err)
+	}
+	var err error
+	defer func() { deleteCgroupOnErr(cgroup, err) }()
```

Comment:
This is one of the few cases where I do suggest a named return param, to be defensive. You got everything right here but it would be very easy for someone to add an `if err := writeCgroupFile(...); err != nil { return err }` later on without noticing that the cleanup wouldn't run
```suggestion
func newJobCgroup(cgroup string, limits Limits) (err error) {
	if err := os.Mkdir(cgroup, 0o750); err != nil {
		return fmt.Errorf("cannot create new job cgroup %q: %w", cgroup, err)
	}
	defer func() { deleteCgroupOnErr(cgroup, err) }()
```

Reply:
Right, I see, this subtle. Fixed 0924a3f